### PR TITLE
Migrate to worker for crown and task completes

### PR DESCRIPTION
**Pull request by austinywang**

### CODE REVIEW

File: apps/server/src/agentSpawner.ts [hunk 1]
```diff
@@ -64,7 +62,7 @@ export async function spawnAgent(
   teamSlugOrId: string
 ): Promise<AgentSpawnResult> {
   try {
-    // Capture the current auth token and header JSON from AsyncLocalStorage so we can
+    // Capture the current auth token and header JSON from AsyncLocalStorages so we can
```
Comment (lawrencecchen) [inline apps/server/src/agentSpawner.ts]: why plural

**Comment by lawrencecchen** [comment]
@codex pls review

File: apps/worker/src/crown/workflow.ts
Comment (lawrencecchen) [inline apps/worker/src/crown/workflow.ts:1]: break down into smaller files

File: apps/worker/src/crown/workflow.ts [hunk 2]
```diff
@@ -0,0 +1,1610 @@
...
+    }
+
+    const prNumber = (() => {
+      if (typeof parsed.number === "number") return parsed.number;
+      if (typeof parsed.number === "string") {
+        const numeric = Number(parsed.number);
+        return Number.isFinite(numeric) ? numeric : undefined;
+      }
+      return undefined;
+    })();
```
Comment (lawrencecchen) [inline apps/worker/src/crown/workflow.ts]: parse with zod

File: apps/worker/src/crown/workflow.ts [hunk 3]
```diff
@@ -0,0 +1,1610 @@
...
+
+export function clearTaskRunContext(taskRunId: string) {
+  taskRunContexts.delete(taskRunId);
+  log("INFO", "Cleared task run context", {
+    taskRunId,
+    remainingContexts: taskRunContexts.size,
+  });
+}
+
+export async function handleWorkerTaskCompletion(
```
Comment (lawrencecchen) [inline apps/worker/src/crown/workflow.ts]: whats taskruncontext for??

File: apps/worker/src/index.ts [hunk 4]
```diff
@@ -827,6 +835,24 @@ async function createTerminal(
...
   } = options;
 
+  const envRecord = env as Record<string, string>;
+  const taskRunToken = envRecord["CMUX_TASK_RUN_JWT"];
+  const convexUrl = envRecord["NEXT_PUBLIC_CONVEX_URL"];
+
+  if (convexUrl) {
+    process.env.NEXT_PUBLIC_CONVEX_URL = convexUrl;
+  }
+
```
Comment (lawrencecchen) [inline apps/worker/src/index.ts:860]: why set this?

File: apps/worker/src/index.ts [hunk 5]
```diff
@@ -827,6 +835,24 @@ async function createTerminal(
     startupCommands = [],
   } = options;
 
+  const envRecord = env as Record<string, string>;
+  const taskRunToken = envRecord["CMUX_TASK_RUN_JWT"];
+  const convexUrl = envRecord["NEXT_PUBLIC_CONVEX_URL"];
+
```
Comment (lawrencecchen) [inline apps/worker/src/index.ts]: is cast necessary?

File: apps/worker/src/index.ts [hunk 6]
```diff
@@ -996,22 +1022,87 @@ async function createTerminal(
...
+            log("INFO", `Crown workflow completed for ${options.taskRunId}`, {
+              taskRunId: options.taskRunId,
+              agentModel: options.agentModel,
+            });
+
+            // Clean up stored exit code
+            if (options.taskRunId) {
+              processExitCodes.delete(options.taskRunId);
+            }
```
Comment (lawrencecchen) [inline apps/worker/src/index.ts]: why don't need taskruncontext, we can just pass to handleWorkerTaskCompletion as an argument

File: apps/worker/src/index.ts [hunk 7]
```diff
@@ -996,22 +1022,87 @@ async function createTerminal(
...
+              elapsedMs: Date.now() - processStartTime,
+            }
+          );
+
+          // Retrieve the exit code if the process has already exited
+          const storedExitCode = processExitCodes.get(options.taskRunId!);
+
+          // Await the crown workflow directly
+          try {
+            await handleWorkerTaskCompletion(options.taskRunId!, {
```
Comment (lawrencecchen) [inline apps/worker/src/index.ts]: dont do !, do if check

File: apps/worker/src/index.ts [hunk 8]
```diff
@@ -1057,11 +1150,23 @@ async function createTerminal(
...
+        log("INFO", `Stored exit code ${exitCode} for task ${options.taskRunId}`);
+      } else if (processExitCodes.delete(options.taskRunId)) {
+        log(
+          "INFO",
+          `Discarded exit code ${exitCode} for task ${options.taskRunId} after workflow cleanup`
+        );
+      }
+    }
+
+    // Still emit to main server for backwards compatibility/logging
```
Comment (lawrencecchen) [inline apps/worker/src/index.ts]: no need for this

File: apps/www/lib/routes/morph.route.ts [hunk 9]
```diff
@@ -269,7 +269,7 @@ morphRouter.openapi(
             if (!existingRepos.has(repoName)) {
               console.log(`Cloning repository: ${repo}`);
 
-              const maxRetries = 3;
+            const maxRetries = 3;
```
Comment (lawrencecchen) [inline apps/www/lib/routes/morph.route.ts]: ? formatting?

File: apps/www/lib/routes/sandboxes.route.ts [hunk 10]
```diff
@@ -183,19 +183,37 @@ sandboxesRouter.openapi(
         return c.text("VSCode or worker service not found", 500);
       }
 
+      // Extract task-related metadata
+      const taskRunId = body.metadata?.taskRunId;
+      const taskRunJwt = body.metadata?.taskRunJwt;
+
```
Comment (lawrencecchen) [inline apps/www/lib/routes/sandboxes.route.ts]: taskRunId and taskRunJwt should NOT be part of metadata, put as part of StartSandboxBody object

File: apps/www/lib/utils/morph-defaults.ts [hunk 11]
```diff
@@ -1,4 +1,11 @@
...
 // export const DEFAULT_MORPH_SNAPSHOT_ID = "snapshot_osr8ngj4";
 // export const DEFAULT_MORPH_SNAPSHOT_ID = "snapshot_r6j3rkmo";
-export const DEFAULT_MORPH_SNAPSHOT_ID = "snapshot_xo8ptbkl";
+// export const DEFAULT_MORPH_SNAPSHOT_ID = "snapshot_xo8ptbkl";
+// export const DEFAULT_MORPH_SNAPSHOT_ID = "snapshot_t7xbu4my";
+// export const DEFAULT_MORPH_SNAPSHOT_ID = "snapshot_qy5fhzlv";
+// export const DEFAULT_MORPH_SNAPSHOT_ID = "snapshot_9rmz7k96";
+// export const DEFAULT_MORPH_SNAPSHOT_ID = "snapshot_bm2228zh";
+// export const DEFAULT_MORPH_SNAPSHOT_ID = "snapshot_6m5g33g6";
+// export const DEFAULT_MORPH_SNAPSHOT_ID = "snapshot_jgwveaed";
```
Comment (lawrencecchen) [inline apps/www/lib/utils/morph-defaults.ts]: rm

File: packages/shared/src/verifyTaskRunToken.ts [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+  token: string
+): Promise<TaskRunTokenPayload> {
+  const verification = await jwtVerify(token, taskRunJwtSecret);
+  const parsed = TaskRunTokenPayloadSchema.safeParse(verification.payload);
+  if (!parsed.success) {
+    throw new Error("Invalid CMUX task run token payload");
+  }
+  return parsed.data;
+}
+
```
Comment (lawrencecchen) [inline packages/shared/src/verifyTaskRunToken.ts:29]: move to shared and rename to verifyTaskRunToken. use in both www and here

File: packages/convex/convex/containerSettings.ts [hunk 13]
```diff
@@ -106,3 +107,29 @@ export const getEffective = authQuery({
...
+      autoCleanupEnabled:
+        settings?.autoCleanupEnabled ?? DEFAULT_SETTINGS.autoCleanupEnabled,
+      stopImmediatelyOnCompletion:
+        settings?.stopImmediatelyOnCompletion ??
+        DEFAULT_SETTINGS.stopImmediatelyOnCompletion,
+      minContainersToKeep:
+        settings?.minContainersToKeep ?? DEFAULT_SETTINGS.minContainersToKeep,
+    };
+  },
+});
```
Comment (lawrencecchen) [inline packages/convex/convex/containerSettings.ts:135]: What's this for? What does crown_http.ts need this?

File: packages/convex/convex/crown_http.ts [hunk 14]
```diff
@@ -15,6 +20,51 @@ const CrownSummarizationRequestSchema = z.object({
...
+  taskId: z.string(),
+  winnerRunId: z.string(),
+  reason: z.string(),
+  evaluationPrompt: z.string(),
+  evaluationResponse: z.string(),
+  candidateRunIds: z.array(z.string()).min(1),
+  summary: z.string().optional(),
+  pullRequest: z
+    .object({
+      url: z.string().url(),
```
Comment (lawrencecchen) [inline packages/convex/convex/crown_http.ts]: use z.url() z.string is deprecated

File: packages/convex/convex/crown_http.ts [hunk 15]
```diff
@@ -127,9 +182,80 @@ async function resolveTeamSlugOrId(
...
+}
+
+async function getOptionalWorkerAuth(
+  req: Request
+): Promise<Response | WorkerAuthContext | null> {
+  const token = req.headers.get("x-cmux-token");
+  if (!token) return null;
+  return await ensureWorkerAuth(req);
+}
+
```
Comment (lawrencecchen) [inline packages/convex/convex/crown_http.ts]: why is function named optional?

File: packages/convex/convex/crown_http.ts [hunk 16]
```diff
@@ -127,9 +182,80 @@ async function resolveTeamSlugOrId(
...
+  try {
+    const payload = await verifyTaskRunToken(token);
+    return { token, payload };
+  } catch (error) {
+    console.error("[convex.crown] Failed to verify task run token", error);
+    return jsonResponse({ code: 401, message: "Unauthorized" }, 401);
+  }
+}
+
```
Comment (lawrencecchen) [inline packages/convex/convex/crown_http.ts]: we can just throw a jsonResponse -- dont need "optional" -- dont return null. or just return null and pass all the way back. dont return a response back.

File: packages/convex/convex/crown_http.ts [hunk 17]
```diff
@@ -189,3 +337,474 @@ export const crownSummarize = httpAction(async (ctx, req) => {
...
+    method: req.method,
+    hasToken: !!req.headers.get("x-cmux-token"),
+  });
+
+  const workerAuth = await ensureWorkerAuth(req);
+  if (workerAuth instanceof Response) return workerAuth;
+
+  const parsed = await ensureJsonRequest(req);
+  if (parsed instanceof Response) return parsed;
+
```
Comment (lawrencecchen) [inline packages/convex/convex/crown_http.ts]: do null checks as outlined above

Comment (lawrencecchen) [reply packages/convex/convex/crown_http.ts]: or just throw

Comment (lawrencecchen) [reply packages/convex/convex/crown_http.ts]: throwing is better

File: packages/convex/convex/crown_http.ts [hunk 18]
```diff
@@ -189,3 +337,474 @@ export const crownSummarize = httpAction(async (ctx, req) => {
...
+        : null,
+      containerSettings: containerSettings
+        ? {
+            autoCleanupEnabled: containerSettings.autoCleanupEnabled,
+            stopImmediatelyOnCompletion:
+              containerSettings.stopImmediatelyOnCompletion,
+            reviewPeriodMinutes: containerSettings.reviewPeriodMinutes,
+          }
+        : null,
+    });
```
Comment (lawrencecchen) [inline packages/convex/convex/crown_http.ts]: why does worker need any of this info?

File: apps/worker/src/crown/workflow.ts [hunk 19]
```diff
@@ -0,0 +1,1610 @@
...
+    reviewPeriodMinutes: number;
+  } | null;
+};
+
+type WorkerAllRunsCompleteResponse = {
+  ok: boolean;
+  taskId: string;
+  allComplete: boolean;
+  statuses: Array<{ id: string; status: string }>;
+};
```
Comment (lawrencecchen) [inline apps/worker/src/crown/workflow.ts]: move to shared dir and share between worker and convex. make convex responses "satisfies" these schemas with typescript satisfies keyword

File: packages/convex/convex/crown_http.ts [hunk 20]
```diff
@@ -189,3 +337,474 @@ export const crownSummarize = httpAction(async (ctx, req) => {
...
+  // Default crown check logic
+  const taskRun = await loadTaskRunForWorker(
+    ctx,
+    workerAuth,
+    validation.data.taskRunId
+  );
+  if (taskRun instanceof Response) return taskRun;
+
+  const taskId = (validation.data.taskId ?? taskRun.taskId) as Id<"tasks">;
+  if (taskId !== taskRun.taskId) {
```
Comment (lawrencecchen) [inline packages/convex/convex/crown_http.ts]: no cast necessary, make validation.data.taskId typedZid in zod schema

File: packages/convex/convex/crown_http.ts [hunk 21]
```diff
@@ -189,3 +337,474 @@ export const crownSummarize = httpAction(async (ctx, req) => {
...
+      userId: workerAuth.payload.userId,
+    }
+  );
+
+  const runsForTeam = runs.filter(
+    (run): run is Doc<"taskRuns"> =>
+      run.teamId === workerAuth.payload.teamId &&
+      run.userId === workerAuth.payload.userId
+  );
+
```
Comment (lawrencecchen) [inline packages/convex/convex/crown_http.ts]: sus. make a new query that returns exactly this.

File: apps/worker/src/crown/workflow.ts [hunk 22]
```diff
@@ -0,0 +1,1610 @@
...
+        hasTaskRun: info?.taskRun,
+      });
+      return;
+    }
+
+    if (info?.taskRun) {
+      runContext.taskId = runContext.taskId ?? info.taskRun.taskId;
+      runContext.teamId = runContext.teamId ?? info.taskRun.teamId;
+    }
+
```
Comment (lawrencecchen) [inline apps/worker/src/crown/workflow.ts]: remove

File: apps/worker/src/crown/workflow.ts [hunk 23]
```diff
@@ -0,0 +1,1610 @@
...
+      maxBuffer: 20 * 1024 * 1024,
+    });
+    const stdout = toUtf8String(result.stdout);
+    return { stdout };
+  } catch (error) {
+    log("ERROR", "Git command failed", { command, error });
+    return null;
+  }
+}
+
```
Comment (lawrencecchen) [inline apps/worker/src/crown/workflow.ts]: why is there runGitCommand AND runGitCommandSafe??

File: apps/worker/src/crown/workflow.ts [hunk 24]
```diff
@@ -0,0 +1,1610 @@
...
+function buildSummarizationPrompt(taskText: string, gitDiff: string): string {
+  return `You are an expert reviewer summarizing a pull request.\n\nGOAL\n- Explain succinctly what changed and why.\n- Call out areas the user should review carefully.\n- Provide a quick test plan to validate the changes.\n\nCONTEXT\n- User's original request:\n${taskText}\n- Relevant diffs (unified):\n${gitDiff || "<no code changes captured>"}\n\nINSTRUCTIONS\n- Base your summary strictly on the provided diffs and request.\n- Be specific about files and functions when possible.\n- Prefer clear bullet points over prose. Keep it under ~300 words.\n- If there are no code changes, say so explicitly and suggest next steps.\n\nOUTPUT FORMAT (Markdown)\n## PR Review Summary\n- What Changed: bullet list\n- Review Focus: bullet list (risks/edge cases)\n- Test Plan: bullet list of practical steps\n- Follow-ups: optional bullets if applicable\n`;
+}
+
```
Comment (lawrencecchen) [inline apps/worker/src/crown/workflow.ts]: why are prompts in worker? prompts should be in convex

File: apps/worker/src/crown/workflow.ts [hunk 25]
```diff
@@ -0,0 +1,1610 @@
...
+        );
+
+        if (!completionState?.ok) {
+          log("ERROR", "Failed to verify task run completion state", {
+            taskRunId,
+            taskId: currentTaskId,
+            attempt,
+          });
+          return;
+        }
```
Comment (lawrencecchen) [inline apps/worker/src/crown/workflow.ts]: funamentally flawed; worker should not determine when all runs are complete. only the server should check if all runs are complete.

File: packages/convex/convex/crown_http.ts [hunk 26]
```diff
@@ -159,28 +295,40 @@ export const crownEvaluate = httpAction(async (ctx, req) => {
...
+  const workerAuthResult = await getOptionalWorkerAuth(req);
+  if (workerAuthResult instanceof Response) return workerAuthResult;
+  const workerAuth = workerAuthResult;
+
+  if (!workerAuth) {
+    const stackAuthError = ensureStackAuth(req);
+    if (stackAuthError) return stackAuthError;
+  }
 
   const parsed = await ensureJsonRequest(req);
```
Comment (lawrencecchen) [inline packages/convex/convex/crown_http.ts]: throw instead

File: packages/convex/convex/crown_http.ts [hunk 27]
```diff
@@ -189,3 +337,474 @@ export const crownSummarize = httpAction(async (ctx, req) => {
...
+  });
+
+  return jsonResponse({
+    ok: true,
+    service: "crown",
+    status: "healthy",
+    timestamp: new Date().toISOString(),
+    version: "1.0.0",
+  });
+});
```
Comment (lawrencecchen) [inline packages/convex/convex/crown_http.ts]: rm

File: packages/convex/convex/crown_http.ts [hunk 28]
```diff
@@ -189,3 +337,474 @@ export const crownSummarize = httpAction(async (ctx, req) => {
...
+  if (existingRun instanceof Response) return existingRun;
+
+  await ctx.runMutation(internal.taskRuns.updateScheduledStopInternal, {
+    taskRunId,
+    scheduledStopAt: validation.data.scheduledStopAt,
+  });
+
+  return jsonResponse({ ok: true });
+});
+
```
Comment (lawrencecchen) [inline packages/convex/convex/crown_http.ts]: this doesnt do anything...

**Review by lawrencecchen — COMMENTED** [review]
fix

**Review by lawrencecchen — CHANGES_REQUESTED** [review]
fix

**Comment by lawrencecchen** [comment]
@codex review

File: apps/worker/src/crown/git.ts [hunk 29]
```diff
@@ -0,0 +1,456 @@
...
+    }
+
+    const { stdout } = await execAsync(
+      `find ${WORKSPACE_ROOT} -maxdepth 2 -type d -name .git 2>/dev/null | head -1`,
+      {
+        cwd: WORKSPACE_ROOT,
+      },
+    );
+
+    if (stdout && stdout.trim()) {
```
Comment (lawrencecchen) [inline apps/worker/src/crown/git.ts]: better to use node fs api instead of execasync here

File: apps/worker/src/crown/git.ts [hunk 30]
```diff
@@ -0,0 +1,456 @@
...
+let gitRepoPath: string | null = null;
+
+export async function detectGitRepoPath(): Promise<string> {
+  if (gitRepoPath) {
+    return gitRepoPath;
+  }
+
+  if (existsSync(join(WORKSPACE_ROOT, ".git"))) {
+    gitRepoPath = WORKSPACE_ROOT;
+    log("INFO", "Git repository found at workspace root", {
```
Comment (lawrencecchen) [inline apps/worker/src/crown/git.ts:22]: - apps/worker/src/crown/git.ts:131 (and later calls at :338, :359, :376, :387, :441) interpolate branch names into shell strings without any quoting. Git allows ', spaces, and other characters in branch names, so a legitimate branch
  like feat'oops will break every one of these commands (and opens the door to shell injection if an attacker controls the branch name). Please either move to spawn/execFile with argument arrays or at least JSON.stringify/quote every
  user-controlled ref before handing it to the shell.
  - apps/worker/src/crown/git.ts:226 converts any failure from cmux-collect-crown-diff.sh into “No changes detected”. When the script bails (e.g. stdout maxBuffer exceeded, missing ref, merge-base failure) the caller gets a clean bill of
  health even though we’ve just lost the diff. That silently hides real errors. We should surface the failure (or at minimum distinguish “script failed” from “no diff”).

**Comment by lawrencecchen** [comment]
@codex check for race conditions and locks and see if it's fine?

File: apps/worker/src/crown/workflow.ts [hunk 31]
```diff
@@ -0,0 +1,636 @@
...
+  );
+
+  if (!info) {
+    log(
+      "ERROR",
+      "Failed to load task run info - endpoint not found or network error",
+      {
+        taskRunId,
+        info,
+        convexUrl: baseUrlOverride || process.env.NEXT_PUBLIC_CONVEX_URL,
```
Comment (lawrencecchen) [inline apps/worker/src/crown/workflow.ts]: get rid of process.env.NEXT_PUBLIC_CONVEX_URL everywhere

File: apps/worker/src/crown/workflow.ts [hunk 32]
```diff
@@ -0,0 +1,636 @@
...
+      hasCompletedRunInfo: Boolean(completedRunInfo),
+      hasInfoTaskRun: Boolean(info?.taskRun),
+    });
+    return;
+  }
+
+  runContext.taskId = realTaskId;
+  runContext.teamId = completedRunInfo.teamId ?? runContext.teamId;
+
+  async function startCrownEvaluation(currentTaskId: string) {
```
Comment (lawrencecchen) [inline apps/worker/src/crown/workflow.ts]: confusing when you declare a fn inside a function, move it out

File: packages/convex/convex/crown_http.ts [hunk 33]
```diff
@@ -127,29 +139,128 @@ async function resolveTeamSlugOrId(
...
+  console.log("[convex.crown] Parsed request body", {
+    bodyType: typeof parsed.json,
+    bodyKeys:
+      parsed.json && typeof parsed.json === "object"
+        ? Object.keys(parsed.json as Record<string, unknown>)
+        : null,
+    body: JSON.stringify(parsed.json),
+  });
+
   const validation = CrownEvaluationRequestSchema.safeParse(parsed.json);
```
Comment (lawrencecchen) [inline packages/convex/convex/crown_http.ts]: rm log

File: packages/convex/convex/crown_http.ts [hunk 34]
```diff
@@ -127,29 +139,128 @@ async function resolveTeamSlugOrId(
...
+
 export const crownEvaluate = httpAction(async (ctx, req) => {
-  const stackAuthError = ensureStackAuth(req);
-  if (stackAuthError) return stackAuthError;
+  console.log("[convex.crown] Evaluate endpoint called", {
+    url: req.url,
+    method: req.method,
+    hasCmuxToken: !!req.headers.get("x-cmux-token"),
+    hasStackAuth: !!req.headers.get("x-stack-auth"),
+  });
```
Comment (lawrencecchen) [inline packages/convex/convex/crown_http.ts]: rm log
